fio-ioengine/flexalloc.c

```diff
@@ -265,8 +313,15 @@ static int fio_flexalloc_init(struct thread_data *td)
 		ret = 1;
 	}
 
-	if (!o->dev_uri) {
-		log_err("flexalloc: dev_uri must be set\n");
+	if (o->daemon_uri && (o->dev_uri || o->md_dev_uri))
```

Comment:
use daemon_mode() here?

Reply:
> > use daemon_mode() here?
> 
> In this case I'd rather not - `daemon_mode()` happens to check if `o->daemon_uri` is set to determine if we are operating in daemon mode or not.
> 
> But at this point we're specifically checking whether certain incompatible parameters are set at the same time - I think it seems clearer this way.

Fair enough

Reply:
> Thanks for looking it over :)
> 
> Regarding point 1 -- nope, there isn't, I haven't fully absorbed the code and have mostly been hacking where it made sense. Is it enforced anywhere else but in the fio file itself? I've tried changing the fio file to read `thread=5` and this has worked alright, even when adding `verify=md5` and `do_verify=1` the jobs complete without any verification errors.

The `thread` option tells fio to use threads for execution instead of (by default) processes. If fio is operating in daemon mode t would be possible to skip these lines:

```
	if (!td->o.use_thread) {
		log_err("flexalloc: thread=1 must be set for the flexalloc ioengine\n");
		ret = 1;
	}
```

However, making this change might be more trouble than it's worth because it will be tricky sorting out what to do when some fio jobs are operating in daemon mode and others are operating in direct mode.

You would want to try `numjobs=5` to have multiple identical jobs running.

> 
> Regarding the compilation: hmm. That sounds like a fio issue, right ?

Yes, the fio compatibility issue is not caused by this PR. There's something not guarded by FIO_IOOPS_VERSION that prevents external ioengines from working.

Reply:
I don't think we should worry about daemon mode vs direct mode. I see this as the same problem of someone using fio on a filesystem and then at the same time running fio on the raw block device underneath the file system. Sure, you can do it, but I don't think anything should be done at the fio level to solve this. I second getting rid of the test that we are using the fio thread option.